Copyright 2020 Andrew M. Olney and made available under [CC BY-SA](https://creativecommons.org/licenses/by-sa/4.0) for text and [Apache-2.0](http://www.apache.org/licenses/LICENSE-2.0) for code.

# TEMPLATE THIS NOTEBOOK TO MAKE FAR CONDITIONS
  
---------------

## Background

We will be working with data on graduate school admissions. The data has four variables:

- `admit`: the admittance status (0=not admitted, 1=admitted)
- `gre`: the student's GRE score
- `gpa`: the student's GPA
- `rank`: rank of the student's undergraduate institution (1=highest to 4=lowest prestige)

In this session, you will solve several problems using this data.

<div class="alert alert-warning">
    
**Condition alternative instructions: Goal structure labels**

### Read CSV into a dataframe
   
</div>

Import the `pandas` library, which lets us work with dataframes.

In [1]:
import pandas as pd

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="i!#]:2XI=^qLb$e.|iwo">pd</variable></variables><block type="importAs" id="8u3elQqk_!6!WoHrlj}e" x="73" y="63"><field name="libraryName">pandas</field><field name="libraryAlias" id="i!#]:2XI=^qLb$e.|iwo">pd</field></block></xml>

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>

Load a dataframe with the data in "datasets/binary.csv" and display it

In [2]:
dataframe = pd.read_csv('datasets/binary.csv')

dataframe

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="t[n^Fcp7,s93E17ZZ9J6">dataframe</variable><variable id="i!#]:2XI=^qLb$e.|iwo">pd</variable></variables><block type="variables_set" id="YMNSvjU:9aS0`rADBobh" x="29" y="215"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field><value name="VALUE"><block type="varDoMethod" id="h[BIiU^0[[vbD`zoBn6+"><field name="VAR" id="i!#]:2XI=^qLb$e.|iwo">pd</field><field name="MEMBER">read_csv</field><data>pd:read_csv</data><value name="INPUT"><block type="text" id="HyH?(x3/MuPXE`T5;)[@"><field name="TEXT">datasets/binary.csv</field></block></value></block></value></block><block type="variables_get" id="uS,Sc{|xLGBqCK`F3*-*" x="8" y="300"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field></block></xml>

admit  gre   gpa  rank
0        0  380  3.61     3
1        1  660  3.67     3
2        1  800  4.00     1
3        1  640  3.19     4
4        0  520  2.93     4
..     ...  ...   ...   ...
395      0  620  4.00     2
396      0  560  3.04     3
397      0  460  2.63     2
398      0  700  3.65     2
399      0  600  3.89     3

[400 rows x 4 columns]

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>

<div class="alert alert-warning">
    
**Condition alternative instructions: Goal structure labels**

### Select rows from a dataframe by position
   
</div>

Show the 3rd row to the final row (total 398 rows).

In [4]:
dataframe[2 : ]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="t[n^Fcp7,s93E17ZZ9J6">dataframe</variable></variables><block type="lists_getSublist" id="5hFiqAq|QPxp%Xc:h/@A" x="8" y="518"><mutation at1="true" at2="false"></mutation><field name="WHERE1">FROM_START</field><field name="WHERE2">LAST</field><value name="LIST"><block type="variables_get" id="Y`nWHnCXvg4!GpZ![Z8S"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field></block></value><value name="AT1"><block type="math_number" id=":;n{~x;%6oD9(oxp_*Pt"><field name="NUM">3</field></block></value></block></xml>

admit  gre   gpa  rank
2        1  800  4.00     1
3        1  640  3.19     4
4        0  520  2.93     4
5        1  760  3.00     2
6        1  560  2.98     1
..     ...  ...   ...   ...
395      0  620  4.00     2
396      0  560  3.04     3
397      0  460  2.63     2
398      0  700  3.65     2
399      0  600  3.89     3

[398 rows x 4 columns]

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>

<div class="alert alert-warning">
    
**Condition alternative instructions: Goal structure labels**

### Select columns from a dataframe by name
   
</div>

Show the last two columns of the data.

In [5]:
dataframe[['gpa', 'rank']]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="t[n^Fcp7,s93E17ZZ9J6">dataframe</variable></variables><block type="indexer" id="oVNd/g7vyxV[^cAbT$JX" x="8" y="300"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field><value name="INDEX"><block type="lists_create_with" id="^oGt7#=i@OMTRwaf:kdd"><mutation items="2"></mutation><value name="ADD0"><block type="text" id="=k__Y{.gxL5z.AJpll.A"><field name="TEXT">gpa</field></block></value><value name="ADD1"><block type="text" id="uQyKHkAg(%zJh!AAX=[#"><field name="TEXT">rank</field></block></value></block></value></block></xml>

gpa  rank
0    3.61     3
1    3.67     3
2    4.00     1
3    3.19     4
4    2.93     4
..    ...   ...
395  4.00     2
396  3.04     3
397  2.63     2
398  3.65     2
399  3.89     3

[400 rows x 2 columns]

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>

<div class="alert alert-warning">
    
**Condition alternative instructions: Goal structure labels**

### Select rows from a dataframe by value
    
</div>

Show the rows where `gpa` is less than 3.

In [10]:
dataframe[(dataframe['gpa'] < 3)]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="t[n^Fcp7,s93E17ZZ9J6">dataframe</variable></variables><block type="indexer" id="Y*eD.jv!V*/5.i:+x([t" x="19" y="390"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field><value name="INDEX"><block type="logic_compare" id="FbNIQ=Q,*rY%1.XvHA^q"><field name="OP">LT</field><value name="A"><block type="indexer" id="d3u5qTTbrdfW-F2:-QR2"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field><value name="INDEX"><block type="text" id="rX][Xdf,hW{pZ4}3u$Sq"><field name="TEXT">gpa</field></block></value></block></value><value name="B"><block type="math_number" id="P*wB;;JrXkeRhwQyoDmg"><field name="NUM">3</field></block></value></block></value></block></xml>

admit  gre   gpa  rank
4        0  520  2.93     4
6        1  560  2.98     1
17       0  360  2.56     3
22       0  600  2.82     4
37       0  520  2.90     3
..     ...  ...   ...   ...
366      0  460  2.93     3
372      1  680  2.42     1
379      0  540  2.70     2
384      1  480  2.62     2
397      0  460  2.63     2

[67 rows x 4 columns]

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>